### PR TITLE
Add Sphinx documentaion for arch, crypto, drivers & interfaces directories

### PR DESCRIPTION
**Pull request by IlievIliya92**
Adding Sphinx documentation for the following directories:

- [arch](https://ilieviliya92.github.io/libcsp/api/arch/arch.html)
- [drivers](https://ilieviliya92.github.io/libcsp/api/drivers/drivers.html)
- [crypto](https://ilieviliya92.github.io/libcsp/api/crypto/crypto.html)
- [interfaces](https://ilieviliya92.github.io/libcsp/api/interfaces/interfaces.html) 

As well a [changelog](https://pypi.org/project/sphinx-changelog/) & [copy_button](https://pypi.org/project/sphinx-copybutton/) extensions have been added with this PR.

### CODE REVIEW

**Comment by yashi** [comment]
I've got the following error when I build:

<details>
<summary>Build log</summary>

```
╭─ yashi@x570 @fd1f9596 ?2 ─────────────────────────── 3s 🐍 3.11.5 libcsp ~/sr/libcsp ─╮
╰─ 23:03:57 ❯ rm -rf build-* && cmake -B build-doc -S doc && cmake --build build-doc   ─╯
CMake Warning (dev) at CMakeLists.txt:1 (project):
  cmake_minimum_required() should be called prior to this top-level project()
  call.  Please see the cmake-commands(7) manual for usage documentation of
  both commands.
This warning is for project developers.  Use -Wno-dev to suppress it.

-- The C compiler identification is GNU 13.2.0
-- The CXX compiler identification is GNU 13.2.0
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: /usr/lib/ccache/cc - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: /usr/lib/ccache/c++ - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Configuring done (0.2s)
-- Generating done (0.0s)
-- Build files have been written to: /home/yashi/src/libcsp/build-doc
[1/1] Building DOCS for lib CSP
FAILED: /home/yashi/src/libcsp/build-docs/index.html 
cd /home/yashi/src/libcsp/build-doc && sphinx-build -E -b html /home/yashi/src/libcsp/doc /home/yashi/src/libcsp/doc/../build-docs /home/yashi/src/libcsp/doc/CI.md /home/yashi/src/libcsp/doc/INSTALL.md /home/yashi/src/libcsp/doc/api/arch/arch.rst /home/yashi/src/libcsp/doc/api/arch/csp_queue_h.rst /home/yashi/src/libcsp/doc/api/arch/csp_time_h.rst /home/yashi/src/libcsp/doc/api/crypto/crypto.rst /home/yashi/src/libcsp/doc/api/crypto/csp_hmac_h.rst /home/yashi/src/libcsp/doc/api/crypto/csp_sha1_h.rst /home/yashi/src/libcsp/doc/api/csp_buffer_h.rst /home/yashi/src/libcsp/doc/api/csp_cmp_h.rst /home/yashi/src/libcsp/doc/api/csp_crc32_h.rst /home/yashi/src/libcsp/doc/api/csp_debug_h.rst /home/yashi/src/libcsp/doc/api/csp_error_h.rst /home/yashi/src/libcsp/doc/api/csp_h.rst /home/yashi/src/libcsp/doc/api/csp_hooks_h.rst /home/yashi/src/libcsp/doc/api/csp_id_h.rst /home/yashi/src/libcsp/doc/api/csp_iflist_h.rst /home/yashi/src/libcsp/doc/api/csp_interface_h.rst /home/yashi/src/libcsp/doc/api/csp_promisc_h.rst /home/yashi/src/libcsp/doc/api/csp_rtable_h.rst /home/yashi/src/libcsp/doc/api/csp_sfp_h.rst /home/yashi/src/libcsp/doc/api/csp_types_h.rst /home/yashi/src/libcsp/doc/api/csp_yaml_h.rst /home/yashi/src/libcsp/doc/api/drivers/can_socketcan_h.rst /home/yashi/src/libcsp/doc/api/drivers/drivers.rst /home/yashi/src/libcsp/doc/api/drivers/eth_linux_h.rst /home/yashi/src/libcsp/doc/api/drivers/usart_h.rst /home/yashi/src/libcsp/doc/api/interfaces/csp_if_can_h.rst /home/yashi/src/libcsp/doc/api/interfaces/csp_if_eth_h.rst /home/yashi/src/libcsp/doc/api/interfaces/csp_if_eth_pbuf_h.rst /home/yashi/src/libcsp/doc/api/interfaces/csp_if_i2c_h.rst /home/yashi/src/libcsp/doc/api/interfaces/csp_if_kiss_h.rst /home/yashi/src/libcsp/doc/api/interfaces/csp_if_lo_h.rst /home/yashi/src/libcsp/doc/api/interfaces/csp_if_tun_h.rst /home/yashi/src/libcsp/doc/api/interfaces/csp_if_udp_h.rst /home/yashi/src/libcsp/doc/api/interfaces/csp_if_zmqhub_h.rst /home/yashi/src/libcsp/doc/api/interfaces/interfaces.rst /home/yashi/src/libcsp/doc/basic.md /home/yashi/src/libcsp/doc/changelog.md /home/yashi/src/libcsp/doc/codestyle.md /home/yashi/src/libcsp/doc/example.md /home/yashi/src/libcsp/doc/history.md /home/yashi/src/libcsp/doc/hooks.md /home/yashi/src/libcsp/doc/index.md /home/yashi/src/libcsp/doc/license.md /home/yashi/src/libcsp/doc/memory.md /home/yashi/src/libcsp/doc/mtu.md /home/yashi/src/libcsp/doc/outflow.md /home/yashi/src/libcsp/doc/protocolstack.md /home/yashi/src/libcsp/doc/structure.md /home/yashi/src/libcsp/doc/topology.md /home/yashi/src/libcsp/doc/tunnel.md
Running Sphinx v4.3.2
WARNING: The config value `version' has type `NoneType', defaults to `str'.
myst v0.18.1: MdParserConfig(commonmark_only=False, gfm_only=False, enable_extensions=[], disable_syntax=[], all_links_external=False, url_schemes=('http', 'https', 'mailto', 'ftp'), ref_domains=None, highlight_code_blocks=True, number_code_blocks=[], title_to_header=False, heading_anchors=None, heading_slug_func=None, footnote_transition=True, words_per_minute=200, sub_delimiters=('{', '}'), linkify_fuzzy_links=True, dmath_allow_labels=True, dmath_allow_space=True, dmath_allow_digits=True, dmath_double_inline=False, update_mathjax=True, mathjax_classes='tex2jax_process|mathjax_process|math|output_area')
building [mo]: targets for 0 po files that are specified
building [html]: 52 source files given on command line
updating environment: [new config] 52 added, 0 changed, 0 removed
reading sources... [  1%] CI
reading sources... [  3%] INSTALL
reading sources... [  5%] api/arch/arch
reading sources... [  7%] api/arch/csp_queue_h
reading sources... [  9%] api/arch/csp_time_h
reading sources... [ 11%] api/crypto/crypto
reading sources... [ 13%] api/crypto/csp_hmac_h
reading sources... [ 15%] api/crypto/csp_sha1_h
/home/yashi/src/libcsp/.venv/lib/python3.11/site-packages/sphinx_c_autodoc/__init__.py:434: RemovedInSphinx50Warning: DocumenterBridge.reporter is deprecated.
  reporter = getattr(self.directive, "reporter", None)
reading sources... [ 17%] api/csp_buffer_h
reading sources... [ 19%] api/csp_cmp_h
reading sources... [ 21%] api/csp_crc32_h
reading sources... [ 23%] api/csp_debug_h
reading sources... [ 25%] api/csp_error_h
reading sources... [ 26%] api/csp_h
reading sources... [ 28%] api/csp_hooks_h
reading sources... [ 30%] api/csp_id_h
reading sources... [ 32%] api/csp_iflist_h
reading sources... [ 34%] api/csp_interface_h
reading sources... [ 36%] api/csp_promisc_h
reading sources... [ 38%] api/csp_rtable_h
reading sources... [ 40%] api/csp_sfp_h
reading sources... [ 42%] api/csp_types_h
reading sources... [ 44%] api/csp_yaml_h
reading sources... [ 46%] api/drivers/can_socketcan_h
reading sources... [ 48%] api/drivers/drivers
reading sources... [ 50%] api/drivers/eth_linux_h
reading sources... [ 51%] api/drivers/usart_h
reading sources... [ 53%] api/interfaces/csp_if_can_h
reading sources... [ 55%] api/interfaces/csp_if_eth_h
reading sources... [ 57%] api/interfaces/csp_if_eth_pbuf_h
reading sources... [ 59%] api/interfaces/csp_if_i2c_h
reading sources... [ 61%] api/interfaces/csp_if_kiss_h
reading sources... [ 63%] api/interfaces/csp_if_lo_h
reading sources... [ 65%] api/interfaces/csp_if_tun_h
reading sources... [ 67%] api/interfaces/csp_if_udp_h
reading sources... [ 69%] api/interfaces/csp_if_zmqhub_h
reading sources... [ 71%] api/interfaces/interfaces
reading sources... [ 73%] basic
reading sources... [ 75%] changelog
reading sources... [ 76%] codestyle
reading sources... [ 78%] example
reading sources... [ 80%] history
reading sources... [ 82%] hooks
reading sources... [ 84%] index
reading sources... [ 86%] license
reading sources... [ 88%] memory
reading sources... [ 90%] mtu
reading sources... [ 92%] outflow
reading sources... [ 94%] protocolstack
reading sources... [ 96%] structure
reading sources... [ 98%] topology
reading sources... [100%] tunnel

looking for now-outdated files... none found
pickling environment... done
checking consistency... done
preparing documents... done
writing output... [  1%] CI
writing output... [  3%] INSTALL
writing output... [  5%] api/arch/arch
writing output... [  7%] api/arch/csp_queue_h
writing output... [  9%] api/arch/csp_time_h
writing output... [ 11%] api/crypto/crypto
writing output... [ 13%] api/crypto/csp_hmac_h
writing output... [ 15%] api/crypto/csp_sha1_h
writing output... [ 17%] api/csp_buffer_h
writing output... [ 19%] api/csp_cmp_h
writing output... [ 21%] api/csp_crc32_h
writing output... [ 23%] api/csp_debug_h
writing output... [ 25%] api/csp_error_h
writing output... [ 26%] api/csp_h
writing output... [ 28%] api/csp_hooks_h
writing output... [ 30%] api/csp_id_h
writing output... [ 32%] api/csp_iflist_h
writing output... [ 34%] api/csp_interface_h
writing output... [ 36%] api/csp_promisc_h
writing output... [ 38%] api/csp_rtable_h
writing output... [ 40%] api/csp_sfp_h
writing output... [ 42%] api/csp_types_h
writing output... [ 44%] api/csp_yaml_h
writing output... [ 46%] api/drivers/can_socketcan_h
writing output... [ 48%] api/drivers/drivers
writing output... [ 50%] api/drivers/eth_linux_h
writing output... [ 51%] api/drivers/usart_h
writing output... [ 53%] api/interfaces/csp_if_can_h
writing output... [ 55%] api/interfaces/csp_if_eth_h
writing output... [ 57%] api/interfaces/csp_if_eth_pbuf_h
writing output... [ 59%] api/interfaces/csp_if_i2c_h
writing output... [ 61%] api/interfaces/csp_if_kiss_h
writing output... [ 63%] api/interfaces/csp_if_lo_h
writing output... [ 65%] api/interfaces/csp_if_tun_h
writing output... [ 67%] api/interfaces/csp_if_udp_h
writing output... [ 69%] api/interfaces/csp_if_zmqhub_h
writing output... [ 71%] api/interfaces/interfaces
writing output... [ 73%] basic
writing output... [ 75%] changelog
writing output... [ 76%] codestyle
writing output... [ 78%] example
writing output... [ 80%] history
writing output... [ 82%] hooks
writing output... [ 84%] index
writing output... [ 86%] license
writing output... [ 88%] memory
writing output... [ 90%] mtu
writing output... [ 92%] outflow
writing output... [ 94%] protocolstack
writing output... [ 96%] structure
writing output... [ 98%] topology
writing output... [100%] tunnel

generating indices... genindex done
highlighting c module code... [  3%] arch/csp_queue.h
highlighting c module code... [  6%] arch/csp_time.h
highlighting c module code... [  9%] crypto/csp_hmac.h
highlighting c module code... [ 12%] crypto/csp_sha1.h
highlighting c module code... [ 16%] csp.h
highlighting c module code... [ 19%] csp_buffer.h
highlighting c module code... [ 22%] csp_cmp.h
highlighting c module code... [ 25%] csp_crc32.h
highlighting c module code... [ 29%] csp_debug.h
highlighting c module code... [ 32%] csp_error.h
highlighting c module code... [ 35%] csp_hooks.h
highlighting c module code... [ 38%] csp_id.h
highlighting c module code... [ 41%] csp_iflist.h
highlighting c module code... [ 45%] csp_interface.h
highlighting c module code... [ 48%] csp_promisc.h
highlighting c module code... [ 51%] csp_rtable.h
highlighting c module code... [ 54%] csp_sfp.h
highlighting c module code... [ 58%] csp_types.h
highlighting c module code... [ 61%] csp_yaml.h
highlighting c module code... [ 64%] drivers/can_socketcan.h
highlighting c module code... [ 67%] drivers/eth_linux.h
highlighting c module code... [ 70%] drivers/usart.h
highlighting c module code... [ 74%] interfaces/csp_if_can.h
highlighting c module code... [ 77%] interfaces/csp_if_eth.h
highlighting c module code... [ 80%] interfaces/csp_if_eth_pbuf.h
highlighting c module code... [ 83%] interfaces/csp_if_i2c.h
highlighting c module code... [ 87%] interfaces/csp_if_kiss.h
highlighting c module code... [ 90%] interfaces/csp_if_lo.h
highlighting c module code... [ 93%] interfaces/csp_if_tun.h
highlighting c module code... [ 96%] interfaces/csp_if_udp.h
highlighting c module code... [100%] interfaces/csp_if_zmqhub.h

writing additional pages... search done
copying static files... done
copying extra files... done
dumping search index in English (code: en)... done
dumping object inventory... failed

Exception occurred:
  File "/usr/lib/python3.11/re/__init__.py", line 185, in sub
    return _compile(pattern, flags).sub(repl, string, count)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
TypeError: expected string or bytes-like object, got 'NoneType'
The full traceback has been saved in /tmp/sphinx-err-tiqmwcac.log, if you want to report the issue to the developers.
Please also report this if it was a user error, so that a better error message can be provided next time.
A bug report can be filed in the tracker at <https://github.com/sphinx-doc/sphinx/issues>. Thanks!
ninja: build stopped: subcommand failed.
╭─ yashi@x570 @fd1f9596 ?2 ─────────────────────────── 3s 🐍 3.11.5 libcsp ~/sr/libcsp ─╮
╰─ 23:04:06 ❯ pip freeze                                                               ─╯
alabaster==0.7.13
Babel==2.13.0
beautifulsoup4==4.12.2
certifi==2023.7.22
charset-normalizer==3.3.0
clang==14.0
docutils==0.17.1
gitdb==4.0.10
GitPython==3.1.37
idna==3.4
imagesize==1.4.1
Jinja2==3.1.2
markdown-it-py==2.2.0
MarkupSafe==2.1.3
mdit-py-plugins==0.3.5
mdurl==0.1.2
myst-parser==0.18.1
packaging==23.2
pbr==5.11.1
Pygments==2.16.1
PyYAML==6.0.1
requests==2.31.0
six==1.16.0
smmap==5.0.1
snowballstemmer==2.2.0
soupsieve==2.5
Sphinx==4.3.2
sphinx-c-autodoc==1.3.0
sphinx-copybutton==0.5.2
sphinx-git==11.0.0
sphinx-rtd-theme==1.3.0
sphinx_design==0.4.1
sphinxcontrib-apidoc==0.3.0
sphinxcontrib-applehelp==1.0.4
sphinxcontrib-devhelp==1.0.2
sphinxcontrib-htmlhelp==2.0.1
sphinxcontrib-jquery==4.1
sphinxcontrib-jsmath==1.0.1
sphinxcontrib-qthelp==1.0.3
sphinxcontrib-serializinghtml==1.1.5
typing_extensions==4.8.0
urllib3==2.0.6
╭─ yashi@x570 @fd1f9596 ?2 ────────────────────────────── 🐍 3.11.5 libcsp ~/sr/libcsp ─╮
╰─ 23:04:21 ❯                                                                          ─╯
```
</details>

**Comment by yashi** [comment]
The word "Refactoring" usually means structual changes.  Your changes seem more like conversions or fixes, at least to me.  Just my two cents.

Oxford Languages:
> restructure (the source code of an application or piece of software) so as to improve operation without altering functionality.

Wikipedia:
> In [computer programming](https://en.wikipedia.org/wiki/Computer_programming) and [software design](https://en.wikipedia.org/wiki/Software_design), code refactoring is the process of restructuring existing [computer code](https://en.wikipedia.org/wiki/Computer_code)—changing the [factoring](https://en.wikipedia.org/wiki/Decomposition_(computer_science))—without changing its external behavior. Refactoring is intended to improve the design, structure, and/or implementation of the [software](https://en.wikipedia.org/wiki/Software) (its [non-functional](https://en.wikipedia.org/wiki/Non-functional_requirement) attributes), while preserving its [functionality](https://en.wikipedia.org/wiki/Functional_requirement).

**Comment by yashi** [comment]
@IlievIliya92 This might be too late but why do you have to convet `@param` and `@returns` to `Parameters:` and `Returns:`? sphinx-c-autodoc seems to be capable of converting the original `@param` format just fine.

**Comment by IlievIliya92** [comment]
> @IlievIliya92 This might be too late but why do you have to convet `@param` and `@returns` to `Parameters:` and `Returns:`? sphinx-c-autodoc seems to be capable of converting the original `@param` format just fine.

This syntax resulted in issues with generating the html docs which is why it was changed.

**Comment by yashi** [comment]
Is it because some header doesn't include `stdint.h`? or any other header files it needs?

**Comment by IlievIliya92** [comment]
> I've got the following error when I build:

I have updated Sphinx to a newer version: 5.3.0

This the build environment from the Docker image:

```
alabaster==0.7.13
Babel==2.12.1
beautifulsoup4==4.12.2
certifi==2023.7.22
charset-normalizer==3.2.0
clang==14.0
docutils==0.17.1
gitdb==4.0.10
GitPython==3.1.37
idna==3.4
imagesize==1.4.1
Jinja2==3.1.2
markdown-it-py==2.2.0
MarkupSafe==2.1.3
mdit-py-plugins==0.3.5
mdurl==0.1.2
myst-parser==0.18.1
packaging==23.1
pbr==5.11.1
Pygments==2.16.1
PyYAML==6.0.1
requests==2.31.0
six==1.16.0
smmap==5.0.1
snowballstemmer==2.2.0
soupsieve==2.5
Sphinx==5.3.0
sphinx-c-autodoc==1.2.2
sphinx-git==11.0.0
sphinx-rtd-theme==1.3.0
sphinx_design==0.4.1
sphinxcontrib-apidoc==0.3.0
sphinxcontrib-applehelp==1.0.4
sphinxcontrib-devhelp==1.0.2
sphinxcontrib-htmlhelp==2.0.1
sphinxcontrib-jquery==4.1
sphinxcontrib-jsmath==1.0.1
sphinxcontrib-qthelp==1.0.3
sphinxcontrib-serializinghtml==1.1.5
typing_extensions==4.8.0
urllib3==2.0.4
```

**Comment by yashi** [comment]
I've got this with the following change.
image

```diff
diff --git a/include/csp/csp_cmp.h b/include/csp/csp_cmp.h
index ab4bc14..75d4f17 100644
--- a/include/csp/csp_cmp.h
+++ b/include/csp/csp_cmp.h
@@ -5,6 +5,7 @@
 #pragma once
 
 #include <csp/csp.h>
+#include <stdint.h>
 
 #ifdef __cplusplus
 extern "C" {
@@ -152,16 +153,13 @@ struct csp_cmp_message {
 /**
  * Generic send management message request.
  *
- * Parameters:
- *  node (uint16_t) [in]: address of subsystem.
- *  timeout (uint32_t) [in]: timeout in mS to wait for reply..
- *  code (uint8_t) [in]: request code.
- *  msg_size (int) [in]: size of \a msg.
- *  msg (struct csp_cmp_message *) [in,out]: data.
- *
- * Returns:
- * int: #CSP_ERR_NONE on success, otherwise an error code.
-*/
+ * @param[in] node address of subsystem.
+ * @param[in] timeout timeout in mS to wait for reply..
+ * @param[in] code request code.
+ * @param[in] msg_size size of \a msg.
+ * @param[in,out] msg data.
+ * @return #CSP_ERR_NONE on success, otherwise an error code.
+ */
 int csp_cmp(uint16_t node, uint32_t timeout, uint8_t code, int msg_size, struct csp_cmp_message *msg);
 
 /**
@@ -181,14 +179,12 @@ CMP_MESSAGE(CSP_CMP_ROUTE_SET_V2, route_set_v2)
 /**
  * Peek (read) memory on remote node.
  *
- * Parameters:
- *	node (uint16_t) [in]: address of subsystem.
- *	timeout (uint32_t) [in]: timeout in mS to wait for reply..
- *	msg (struct csp_cmp_message *) [in/out]: memory address and number of bytes to peek. (msg peeked/read memory)
- *
- * Returns:
- * 	int: #CSP_ERR_NONE on success, otherwise an error code.
- */
+ * @param node[in]: address of subsystem.
+ * @param timeout[in]: timeout in mS to wait for reply..
+ * @param msg[in]: memory address and number of bytes to peek.
+ * @param msg[out]: peeked/read memory.
+ * @return #CSP_ERR_NONE on success, otherwise an error code.
+*/
 static inline int csp_cmp_peek(uint16_t node, uint32_t timeout, struct csp_cmp_message *msg) {
 	return csp_cmp(node, timeout, CSP_CMP_PEEK, CMP_SIZE(peek) - sizeof(msg->peek.data) + msg->peek.len, msg);
 }
@@ -196,13 +192,11 @@ static inline int csp_cmp_peek(uint16_t node, uint32_t timeout, struct csp_cmp_m
 /**
  * Poke (write) memory on remote node.
  *
- * Parameters:
- *	node (uint16_t) [in]: address of subsystem.
- *	timeout (uint32_t) [in]: timeout in mS to wait for reply..
- *	msg (struct csp_cmp_message *) [in]: memory address, number of bytes and the actual bytes to poke/write.
+ * @param node[in]: address of subsystem.
+ * @param timeout[in]: timeout in mS to wait for reply..
+ * @param msg[in]: memory address, number of bytes and the actual bytes to poke/write.
  *
- * Returns:
- *	int: #CSP_ERR_NONE on success, otherwise an error code.
+ * @return #CSP_ERR_NONE on success, otherwise an error code.
  */
 static inline int csp_cmp_poke(uint16_t node, uint32_t timeout, struct csp_cmp_message *msg) {
 	return csp_cmp(node, timeout, CSP_CMP_POKE, CMP_SIZE(poke) - sizeof(msg->poke.data) + msg->poke.len, msg);
```

**Comment by IlievIliya92** [comment]
> I've got this with the following change. 

That's nice! I guess it was an issue on my side. Although this comment formatting seems to not display the data types of the input parameters. I guess we can change it back to `@param` & `@returns` syntax if needed.

I was wondering what happened with this Action `Build and Test / run-test (ubuntu-22.04, posix, meson, gcc, g++) ` do you have any suggestion what could be wrong here? I think this job passed on my [local branch](https://github.com/IlievIliya92/libcsp/actions/runs/6415147601/job/17416592852)

**Comment by yashi** [comment]
> I was wondering what happened with this Action Build and Test / run-test (ubuntu-22.04, posix, meson, gcc, g++) 

I've re-run the test and it succeeded.

I've seen the same thing with my PR, too.  It seems ping was timed out.  I assume the backend was a bit busy and did not schedule it for a second. (I know one second is loooong time for schedulers).

**Comment by yashi** [comment]
> Although this comment formatting seems to not display the data types of the input parameters.

Right. But do we really need type info in the parameter lists? We have the function signature right above, and parameter types are clickable in the signature line, too. I'd say it's better for maintainability (a single place to fix) without compromising usability. What do you think?

**Comment by IlievIliya92** [comment]
> I'd say it's better for maintainability (a single place to fix)

Using the `Parameters` notation gives more explicit overview of the data types but  I agree it is always better to make the changes at one place.

I was having issues with that syntax because of the missing header `#include <stdint.h>`.  Including this header will be required in most of the files.

Should  we agree to use the `@param`  & `@return` syntax ?

**Comment by yashi** [comment]
> Should we agree to use the @param & @return syntax ?

Sure, I'm fine with that.  (It might be because I'm used to write JavaDoc / Doxygen style)

Just for a reference: https://www.doxygen.nl/manual/commands.html#cmdparam

**Comment by yashi** [comment]
params should be 
```
@param[in] msg Description
```
Instead of
```
@param msg[in] Description
```
My patch has wrong orders. And a colon `:` is not required.

**Comment by yashi** [comment]
> because of the missing header #include <stdint.h>. Including this header will be required in most of the files.

Actually, we should fix it anyway to make our header files [self-contained](https://google.github.io/styleguide/cppguide.html#Self_contained_Headers).

**Comment by yashi** [comment]
I still can't build it

```
writing additional pages... search done
copying static files... done
copying extra files... done
dumping search index in English (code: en)... done
dumping object inventory... failed

Exception occurred:
  File "/usr/lib/python3.11/re/__init__.py", line 185, in sub
    return _compile(pattern, flags).sub(repl, string, count)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
TypeError: expected string or bytes-like object, got 'NoneType'
The full traceback has been saved in /tmp/sphinx-err-gx_cfud_.log, if you want to report the issue to the developers.
Please also report this if it was a user error, so that a better error message can be provided next time.
A bug report can be filed in the tracker at <https://github.com/sphinx-doc/sphinx/issues>. Thanks!
ninja: build stopped: subcommand failed.
```

```

❯ pip freeze
alabaster==0.7.13
Babel==2.13.0
beautifulsoup4==4.12.2
certifi==2023.7.22
charset-normalizer==3.3.0
clang==14.0
docutils==0.18.1
gitdb==4.0.10
GitPython==3.1.37
idna==3.4
imagesize==1.4.1
Jinja2==3.1.2
markdown-it-py==2.2.0
MarkupSafe==2.1.3
mdit-py-plugins==0.3.5
mdurl==0.1.2
myst-parser==1.0.0
packaging==23.2
pbr==5.11.1
Pygments==2.16.1
PyYAML==6.0.1
requests==2.31.0
six==1.16.0
smmap==5.0.1
snowballstemmer==2.2.0
soupsieve==2.5
Sphinx==5.3.0
sphinx-c-autodoc==1.3.0
sphinx-copybutton==0.5.2
sphinx-git==11.0.0
sphinx-rtd-theme==1.3.0
sphinx_design==0.5.0
sphinxcontrib-apidoc==0.4.0
sphinxcontrib-applehelp==1.0.7
sphinxcontrib-devhelp==1.0.5
sphinxcontrib-htmlhelp==2.0.4
sphinxcontrib-jquery==4.1
sphinxcontrib-jsmath==1.0.1
sphinxcontrib-qthelp==1.0.6
sphinxcontrib-serializinghtml==1.1.9
urllib3==2.0.6
```

Attaching [sphinx-err-gx_cfud_.log](https://github.com/libcsp/libcsp/files/12840367/sphinx-err-gx_cfud_.log)

**Comment by IlievIliya92** [comment]
I have tried to reproduce the same environment using a Docker image based on Ubuntu 23.04 and the build seems to be passing well.

**Dockerfile**

```
FROM ubuntu:23.04
ARG DEBIAN_FRONTEND=noninteractive

# Install dependencies
RUN apt-get update && \
    apt-get install --no-install-recommends --no-install-suggests -y sudo git python3-pip cmake \
    libclang-14-dev build-essential locales locales-all ninja-build

# Install libxdp & libbpf
RUN mkdir /home/libcsp_sphinx_docs && \
    cd /home/libcsp_sphinx_docs && \
    git clone https://github.com/IlievIliya92/libcsp.git --branch develop && \
    cd libcsp/doc && \
    pip3 install --break-system-packages -r requirements.txt && \
    mkdir build && \
    cd build && \
    cmake -GNinja .. && \
    ninja

ENV LC_ALL en_US.UTF-8
ENV LANG en_US.utf8
ENV LANGUAGE en_US.UTF-8

# Set working directory
WORKDIR /home/libcsp_sphinx_docs
CMD /bin/bash
```

The following is the output of the `pip freeze` command from the Docker image:

```
pip freeze
alabaster==0.7.13
Babel==2.13.0
beautifulsoup4==4.12.2
certifi==2023.7.22
charset-normalizer==3.3.0
clang==14.0
docutils==0.18.1
gitdb==4.0.10
GitPython==3.1.37
idna==3.4
imagesize==1.4.1
Jinja2==3.1.2
markdown-it-py==2.2.0
MarkupSafe==2.1.3
mdit-py-plugins==0.3.5
mdurl==0.1.2
myst-parser==1.0.0
packaging==23.2
pbr==5.11.1
Pygments==2.16.1
PyYAML==6.0.1
requests==2.31.0
six==1.16.0
smmap==5.0.1
snowballstemmer==2.2.0
soupsieve==2.5
Sphinx==5.3.0
sphinx-c-autodoc==1.3.0
sphinx-copybutton==0.5.2
sphinx-git==11.0.0
sphinx-rtd-theme==1.3.0
sphinx_design==0.5.0
sphinxcontrib-apidoc==0.4.0
sphinxcontrib-applehelp==1.0.7
sphinxcontrib-devhelp==1.0.5
sphinxcontrib-htmlhelp==2.0.4
sphinxcontrib-jquery==4.1
sphinxcontrib-jsmath==1.0.1
sphinxcontrib-qthelp==1.0.6
sphinxcontrib-serializinghtml==1.1.9
urllib3==2.0.6
```

The only difference I have noticed is the version of Python3.11. The build in the Docker container uses `Python 3.11.4` and from the ` sphinx-err-gx_cfud_.log` I've seen it is `Python 3.11.5`. I am really not able to  reproduce this issue on my end. I really haven't tried the build in virtual environment. I am wondering if that could be the issue.

**Comment by yashi** [comment]
OK.  I've found out.  I've been testing your PR on detached head.  And `get_active_branch_name()` returns `None` on it.  I've queued up the following change. If you like it, I'll merge it after yours.

```diff
diff --git a/doc/conf.py b/doc/conf.py
index a6ef4ee..724d811 100644
--- a/doc/conf.py
+++ b/doc/conf.py
@@ -6,6 +6,7 @@
 
 # -- Imports -----------------------------------------------------------------
 from pathlib import Path
+from pygit2 import Repository
 import clang.cindex as cl
 
 # -- Path setup --------------------------------------------------------------
@@ -69,15 +70,7 @@ html_theme_options = {
     'sticky_navigation': True
 }
 
-def get_active_branch_name():
-    head_dir = Path("../") / ".git" / "HEAD"
-    with head_dir.open("r") as f:
-        content = f.read().splitlines()
-
-    for line in content:
-        if line[0:4] == "ref:":
-            return line.partition("refs/heads/")[2]
-version = get_active_branch_name()
+version = Repository('.').head.shorthand
 
 # Add any paths that contain custom static files (such as style sheets) here,
 # relative to this directory. They are copied after the builtin static files,
diff --git a/doc/requirements.txt b/doc/requirements.txt
index c16e8b7..9b8bba8 100644
--- a/doc/requirements.txt
+++ b/doc/requirements.txt
@@ -7,3 +7,4 @@ sphinx-rtd-theme
 sphinx_design
 sphinx_git
 sphinx_copybutton
+pygit2

```

**Comment by yashi** [comment]
@IlievIliya92 Please make a dedicated commit for the header include change.  And please describe why on the commit.  The reason is that the include change is non-documentation change and we don't want it to be burried in the others. 
 Thanks.

**Comment by yashi** [comment]
@IlievIliya92 Changing `#include <>` to `#include ""` is not the way we want to take it.  Instead, tell sphinx-c-autodoc where to find include files by setting [`c_autodoc_compilation_args`](https://sphinx-c-autodoc.readthedocs.io/en/latest/configuration.html#c-autodoc-compilation-args)

```diff
diff --git a/doc/conf.py b/doc/conf.py
index 91cdf01..470f444 100644
--- a/doc/conf.py
+++ b/doc/conf.py
@@ -44,6 +44,7 @@ exclude_patterns = []
 
 # -- API C files -------------------------------------------------------------
 c_autodoc_roots = ['../include/csp']
+c_autodoc_compilation_args = ["-I../src", "-I../include"]
 
 
 # -- Options for HTML output -------------------------------------------------

```

You might also need `builddir/include` for `csp/autoconfig.h`.  But it seems many (if not all) headers are fine without it.  (Meaning, I don't want to hard code `builddir`)

I've also tried to use [`c_autodoc_compilation_database`](https://sphinx-c-autodoc.readthedocs.io/en/latest/configuration.html#c-autodoc-compilation-database) but failed.

**Comment by IlievIliya92** [comment]
> Instead, tell sphinx-c-autodoc where to find include files

Thanks! That is very elegant solution.

File: doc/conf.py [hunk 1]
```diff
@@ -44,7 +44,7 @@
 
 # -- API C files -------------------------------------------------------------
 c_autodoc_roots = ['../include/csp']
-
+c_autodoc_compilation_args = ["-I./../../include/"]
```
Comment (yashi) [inline doc/conf.py]: Are you sure about this path? I don't get right output with the current settings.
image

**Review by yashi — CHANGES_REQUESTED** [review]

File: doc/conf.py (hunk 1 above)
Comment (IlievIliya92) [reply doc/conf.py]: I have made the changes and fix the include path.

Building the docs with:

```
cmake -B build-docs -S doc
cmake --build build-docs
```

should result in a properly formatted HTML.

**Review by yashi — APPROVED** [review]

**Pull request merged**